src/Native/Runtime/amd64/PInvoke.asm

```diff
@@ -341,4 +340,59 @@ LEAF_ENTRY RhpReversePInvokeReturn, _TEXT
 LEAF_END RhpReversePInvokeReturn, _TEXT
 
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; RhpPInvoke
+;;
+;; IN:  RCX: address of pinvoke frame
+;;
+;; TRASHES: R10, R11
+;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+LEAF_ENTRY RhpPInvoke, _TEXT
+        ;; R10 = GetThread(), TRASHES R11
+        INLINE_GETTHREAD r10, r11
+
+        ;; The callsite of the helper is as good to start the stackwalk as the actual PInvoke callsite
+        mov         r11, [rsp]                  ; r11 <- return address
+        mov         qword ptr [rcx + OFFSETOF__PInvokeTransitionFrame__m_pThread], r10
+        mov         qword ptr [rcx + OFFSETOF__PInvokeTransitionFrame__m_FramePointer], rbp
+        mov         qword ptr [rcx + OFFSETOF__PInvokeTransitionFrame__m_RIP], r11
+
+        lea         r11, [rsp + 8]              ; r11 <- caller SP
+        mov         dword ptr [rcx + OFFSETOF__PInvokeTransitionFrame__m_dwFlags], PTFF_SAVE_RSP
+        mov         qword ptr [rcx + OFFSETOF__PInvokeTransitionFrame__m_PreservedRegs], r11
+
+        mov         qword ptr [r10 + OFFSETOF__Thread__m_pTransitionFrame], rcx
+
+        cmp         [RhpTrapThreads], 0 
+        jne         @F                  ; forward branch - predicted not taken
+        ret
+@@:
+        jmp         RhpWaitForSuspend
+
+        ret
+LEAF_END RhpPInvoke, _TEXT
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;
+;; RhpPInvokeReturn
+;;
+;; IN:  RCX: address of pinvoke frame
+;;
+;; TRASHES: RCX, RDX, R8, R9, R10, R11
+;; 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+LEAF_ENTRY RhpPInvokeReturn, _TEXT
+        mov         rdx, [rcx + OFFSETOF__PInvokeTransitionFrame__m_pThread]
+        mov         qword ptr [rdx + OFFSETOF__Thread__m_pTransitionFrame], 0
+        cmp         [RhpTrapThreads], 0
+        jne         @f                  ; forward branch - predicted not taken
+        ret
+@@:
+        jmp         RhpWaitForGC
```

Comment:
Please add a comment to the effect that rcx still contains the transition frame, as this is an argument to RhpWaitForGC